index.js

```diff
@@ -34,24 +34,68 @@ function focusTrap(element, userOptions) {
 
   return trap;
 
+  function delayFocusTrapActivation(checkCanActivate, callback) {
+    var startDate = Date.now();
+    var interval = setInterval(function() {
+      if (checkCanActivate(container)) {
+        clearInterval(interval);
+        callback();
+      } else {
+        var timeDifferenceInSeconds = (Date.now() - startDate) / 1000;
+        if (timeDifferenceInSeconds > 10) {
+          // eslint-disable-next-line no-console
+          console.warn(
+            [
+              'Focus-Trap activation for the following element timed out after',
+              timeDifferenceInSeconds,
+              'seconds'
+            ].join(' '),
+            container
+          );
+          clearInterval(interval);
+        }
+      }
+    }, 5);
+  }
+
   function activate(activateOptions) {
     if (state.active) return;
 
-    updateTabbableNodes();
-
     state.active = true;
     state.paused = false;
     state.nodeFocusedBeforeActivation = doc.activeElement;
 
-    var onActivate =
-      activateOptions && activateOptions.onActivate
-        ? activateOptions.onActivate
-        : config.onActivate;
-    if (onActivate) {
-      onActivate();
+    function getOption(optionName) {
+      return activateOptions && activateOptions[optionName]
+        ? activateOptions[optionName]
+        : config[optionName];
+    }
+
+    var onActivate = getOption('onActivate');
+    var onSuccessfulActivation = getOption('onSuccessfulActivation');
+    var checkCanActivate = getOption('checkCanActivate');
+
+    if (checkCanActivate) {
+      if (onActivate) {
+        onActivate();
+      }
+
+      delayFocusTrapActivation(checkCanActivate, function() {
+        updateTabbableNodes();
+        addListeners();
+        if (onSuccessfulActivation) {
+          onSuccessfulActivation();
+        }
+      });
+    } else {
+      updateTabbableNodes();
+      if (onActivate) {
+        onActivate();
```

Comment:
I would like to call `onActivate` before `updateTabbableNodes` here to be consistent with the `checkCanActivate` version but that could potentially break existing code so it needs to come after `updateTabbableNodes()`